### PR TITLE
The mostly agreed parts of the evidence-verification-min PR

### PR DESCRIPTION
**Pull request by andrew-draper**

### CODE REVIEW

File: draft-ietf-rats-corim.md [hunk 1]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+* Appraisal Context initialisation
+* Evidence collection
+* Evidence appraisal
+
+At a few well-defined points in the procedure, the Verifier behaviour will
+depend on the specific CoRIM profile.
+Each CoRIM profile MUST provide a description of the expected Verifier behavior
+for each of those well-defined points.
+
+Note that what follows describes a simplified implementation.
```
Comment (yogeshbdeshpande) [inline draft-ietf-rats-corim.md:1223]:
```suggestion
Note that what follows describes a simplified algorithm for verification.
```

File: draft-ietf-rats-corim.md [hunk 2]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+Other selection criteria MAY be applied.
+
+For example, if the Evidence format is known in advance, tags that do not match
+the expected profile can be readily discarded.
+
+The selection process MUST yield at least one usable tag.
+
+### CoBOM Extraction
+
+All the available Concise Bill Of Material (CoBOMs) tags are then collected
```
Comment (yogeshbdeshpande) [inline draft-ietf-rats-corim.md:1261]: Can you please help me clarify:

For a given profile, only one active CoBOM is expected at the point of verification.

So does this section assumes that the set of CoBOMs pertain to all the CoBOMs (for a variety of attester types) and thus all profiles which Verifier handles?
If such is the assumption, then one must explain how to filter the `active` CoBOM relevant to a specific profile.

File: draft-ietf-rats-corim.md [hunk 3]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+
+After the Verifier has processed all CoBOMs it MUST discard any tags which have
+not been activated by a CoBOM.
+
+### Tags Identification and Validation
+
+The Verifier chooses tags -- including Concise Module ID Tags (CoMID, {{sec-comid}}),
+Concise Software ID Tags (CoSWID, {{-coswid}}),
+and/or Concise Trust Anchor Stores (CoTS, {{?I-D.ietf-rats-concise-ta-stores}}) --
+from the selected CoRIMs.
```
Comment (yogeshbdeshpande) [inline draft-ietf-rats-corim.md:1274]:
```suggestion
from the selected CoRIMs activated by a CoBOM.
```

File: draft-ietf-rats-corim.md [hunk 4]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+
+The Accepted Claims Set will be matched against CoMID reference values, as per
+the appraisal policy of the Verifier.
+This document describes an example evidence structure which can be easily
+matched against these reference values.
+Each set of evidence contains an `environment-map` providing a namespace, and
+a non empty `measurement-values-map`.
+
+Each entry in the `measurement-values-map` is a separate piece of evidence
+describing the environment named by the `environment-map`.
```
Comment (yogeshbdeshpande) [inline draft-ietf-rats-corim.md:1357]:
```suggestion
describing the measurements associated to the environment identified in the `environment-map`.
```

File: draft-ietf-rats-corim.md [hunk 5]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+
+ An Attester can provide multiple `environment-map`s each containing a
+ `measurement-values-map` with one entry;  a single `environment-map` containing
```
Comment (yogeshbdeshpande) [inline draft-ietf-rats-corim.md:1360]: 100 % agree with statement on line 1360 and hence we should support this expression natively in the base CDDL.

File: draft-ietf-rats-corim.md [hunk 6]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+ multiple entries in its `measurement-values-map`; or a combination of
+ these approaches.
+
+If evidence from different sources has the same `environment-map` then the
+`measurement-values-map`s are merged.
+If both measurement-value-maps being merged contain the same key then the
```
Comment (yogeshbdeshpande) [inline draft-ietf-rats-corim.md:1366]:
```suggestion
If measurement-value-maps being merged contain the same key then the
```

File: draft-ietf-rats-corim.md [hunk 7]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+values associated with that key MUST be binary identical.
```
Comment (yogeshbdeshpande) [inline draft-ietf-rats-corim.md:1367]: Also can you please clarify what you meant by `key` here? is this referring to a key-value pair concept or something else?

File: draft-ietf-rats-corim.md [hunk 8]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+> A CoRIM profile MUST describe:
+>
+> * How cryptographic verification key material is represented (e.g., using Attestation Keys triples, or CoTS tags)
+> * How key material is associated with the Attesting Environment
+> * How the Attesting Environment is identified in Evidence
+
+### The Accepted Claims Set
+
+At the end of the Evidence collection process evidence has been converted into
+a format suitable for appraisal. This document describes an `accepted-claims-set`
```
Comment (yogeshbdeshpande) [inline draft-ietf-rats-corim.md:1334]:
```suggestion
a format suitable for appraisal, known as `accepted-claims-set`. This document describes an `accepted-claims-set`
```

**Review by yogeshbdeshpande — APPROVED** [review]
Hi @andrew-draper
Thanks for the Pull Request. The text seems to be in-line with what we have discussed so far.

I have left some Nits for you to look at but the principle looks to be in good shape!

File: draft-ietf-rats-corim.md (hunk 3 above)
Comment (thomas-fossati) [reply draft-ietf-rats-corim.md:1274]: @yogeshbdeshpande: the text you suggest seems redundant given we are past the CoBOM processing step.  (But see also my proposal to merge the subsections.)

File: draft-ietf-rats-corim.md [hunk 9]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+The Verifier MUST discard all tags which are not syntactically and semantically
+valid.
+In particular, any cross-referenced triples (e.g., CoMID-CoSWID linking triples)
+MUST be successfully resolved.
+
+### Appraisal Context Construction
+
+All of the validated and potentially useful tags are loaded into the Appraisal Context.
+Each tag is loaded into the Appraisal Context as a single unit, later stages
+will accept or ignore the whole tag.
```
Comment (thomas-fossati) [inline draft-ietf-rats-corim.md:1285]: unclear what this means.  Is there a typo?

File: draft-ietf-rats-corim.md [hunk 10]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+
+The exact cryptographic signature validation mechanics depend on the specific
+Evidence collection protocol.
+
+For example:
+In DICE, a proof of liveness is performed on the final key in the certificate
+chain.
+If this passes then a suitable certification path anchored on a trusted root
+certificate is looked up -- e.g., based on linking information obtained from
+the DeviceID certificate (see Section 9.2.1 of {{DICE-Layering-Architecture}}) --
```
Comment (thomas-fossati) [inline draft-ietf-rats-corim.md:1315]: remember to add the bib entry for DICE-Layering-Architecture or it won't compile.

File: draft-ietf-rats-corim.md [hunk 11]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+in the Appraisal Context.  If found, then usual X.509 certificate validation
+is performed.
+In PSA, the verification public key is looked up in the appraisal context using
+the `euid` claim found in the PSA claims-set (see {{Section 4.2.1 of -psa-token}}).
```
Comment (thomas-fossati) [inline draft-ietf-rats-corim.md:1319]:
```suggestion
the `ueid` claim found in the PSA claims-set (see {{Section 4.2.1 of -psa-token}}).
```

File: draft-ietf-rats-corim.md (hunk 8 above)
Comment (thomas-fossati) [reply draft-ietf-rats-corim.md:1334]: I don't think the specific suggestion works (esp. in the flow with the rest of the sentence).  But I agree we need to be a bit more precise here about what the `accepted-claims-set` is supposed to represent in our "exemplary" verification flow.  Maybe:
```suggestion
a format suitable for appraisal. To this end, this document describes an `accepted-claims-set`
```

File: draft-ietf-rats-corim.md [hunk 12]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+The Verifier MUST activate all tags referenced by a CoBOM.
+
+After the Verifier has processed all CoBOMs it MUST discard any tags which have
+not been activated by a CoBOM.
+
+### Tags Identification and Validation
+
+The Verifier chooses tags -- including Concise Module ID Tags (CoMID, {{sec-comid}}),
+Concise Software ID Tags (CoSWID, {{-coswid}}),
+and/or Concise Trust Anchor Stores (CoTS, {{?I-D.ietf-rats-concise-ta-stores}}) --
```
Comment (thomas-fossati) [inline draft-ietf-rats-corim.md:1273]: use the already defined alias (`ta-store`) or it will break:

```suggestion
and/or Concise Trust Anchor Stores (CoTS, {{-ta-store}}) --
```

File: draft-ietf-rats-corim.md [hunk 13]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+format and the algorithms used to compare it against CoMID reference values.
+
+    accepted-claims-set = {​
+        &(ce.evidence-triples: 0) => [ + reference-triple-record ]​
+      ? &(ce.identity-triples: 1) => [ + identity-triple-record ]​
+      ? &(ce.dependency-triples: 2) => [ + domain-dependency-triple-record ]
+      ? &(ce.domain-membership-triples: 3) => [ + domain-membership-triple-record ]
+      ? &(ce.coswid-triples: 4) => [ + ev-coswid-triple-record ]
+      * $$ev-triples-map-extension​
+    }
```
Comment (thomas-fossati) [inline draft-ietf-rats-corim.md:1344]: we probably want to move this together with the rest of the CDDL (to apply automatic validation & consistency checks).  If you'd rather make it in a separate PR, please raise an issue from this comment.

File: draft-ietf-rats-corim.md [hunk 14]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+### Accepted Claims Set Initialisation
+
+The Accepted Claims Set is initialised with cryptographically verified Evidence
+from the Attestation Environments.
+
+> A CoRIM profile MUST describe:
+>
+> * How evidence is converted to a format suitable for appraisal
+
+Section {sec-dice-spdm} provides information on how evidence collected using
```
Comment (thomas-fossati) [inline draft-ietf-rats-corim.md:1378]: refs are `{{...}}`.  Besides, the processor is smart enough to prepend Section, Figure depending on the anchor.  So what you need is:
```suggestion
{{sec-dice-spdm}} provides information on how evidence collected using
```

File: draft-ietf-rats-corim.md [hunk 15]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+
+CoRIM that are secured by a cryptographic mechanism such as a signature which
+does not pass validation MUST be discarded.
+
+Other selection criteria MAY be applied.
+
+For example, if the Evidence format is known in advance, tags that do not match
+the expected profile can be readily discarded.
+
+The selection process MUST yield at least one usable tag.
```
[the remark this answers is not in the document]
Comment (nedmsmith) [reply draft-ietf-rats-corim.md:1257]: At this point, the RIM is discarded (except that the RIM's signer is remembered - possibly by injecting `authorized-by` attributes). The next level processing applies to tags. The current text seems correct.

File: draft-ietf-rats-corim.md (hunk 9 above)
Comment (nedmsmith) [reply draft-ietf-rats-corim.md:1285]: possibly: "...will accept or ignore the tag in its entirety."

File: draft-ietf-rats-corim.md (hunk 1 above)
Comment (nedmsmith) [reply draft-ietf-rats-corim.md:1223]: Note that what follows describes a simplified algorithm.

File: draft-ietf-rats-corim.md (hunk 10 above)
Comment (nedmsmith) [reply draft-ietf-rats-corim.md:1315]: DICE Layering Architecture, Version 1.0, Revision 0.19, July 23, 2020. https://trustedcomputinggroup.org/wp-content/uploads/DICE-Layering-Architecture-r19_pub.pdf

File: draft-ietf-rats-corim.md (hunk 2 above)
Comment (nedmsmith) [reply draft-ietf-rats-corim.md:1261]: CoBOMs can contain tags that are not in the RIM that immediately contains the CoBOM. Dependent RIMS (containing these tags) may, themselves, have CoBOMs that reference tags in the immediate RIM (or yet other RIMS), and so forth. 
Presumably, there is a "top level" CoBOM that correctly describes the "device" that is being attested. The question for Verifiers is how does it know it has the complete description (via tags) of the device? Which CoBOM is the 'root' CoBOM?

File: draft-ietf-rats-corim.md [hunk 16]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+The goal of the initialisation phase is to load the CoRIM Appraisal Context
+with objects such as tags (CoMID, CoSWID, etc.) from CoRIM files,
+cryptographic validation key material (e.g., raw public keys, root certificates,
+intermediate CA certificate chains), etc. that will be used in the subsequent
+Evidence Appraisal phase.
+
+### CoRIM Selection
+
+All available CoRIMs are collected.
+A Verifier may be pre-configured with a large number of tags describing many
```
Comment (yogeshbdeshpande) [inline draft-ietf-rats-corim.md:1241]:
```suggestion
A Verifier may be pre-configured with a large number of CoRIMs describing many
```

File: draft-ietf-rats-corim.md [hunk 17]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+
+CoRIMs that are not within their validity period, or that cannot be associated
+with an authenticated and authorised source MUST be discarded.
+
+CoRIM that are secured by a cryptographic mechanism such as a signature which
+does not pass validation MUST be discarded.
+
+Other selection criteria MAY be applied.
+
+For example, if the Evidence format is known in advance, tags that do not match
```
Comment (yogeshbdeshpande) [inline draft-ietf-rats-corim.md:1254]:
```suggestion
For example, if the Evidence format is known in advance, CoRIMs that do not match
```

File: draft-ietf-rats-corim.md [hunk 18]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+the expected profile can be readily discarded.
```
Comment (yogeshbdeshpande) [inline draft-ietf-rats-corim.md:1255]:
```suggestion
For example, if the Evidence format is known in advance, CoRIMs using a profile that is not understood by a Verifier can be readily discarded.
```

Comment (nedmsmith) [reply draft-ietf-rats-corim.md:1255]: Given matching semantics are encoded by triples, shouldn't the language here focus on triples?

File: draft-ietf-rats-corim.md (hunk 1 above)
Comment (yogeshbdeshpande) [reply draft-ietf-rats-corim.md:1230]: @nedmsmith : Me and Andrew thinks, "standard algorithm is neat and is specified in the standard document, so works for us atleast!

Comment (nedmsmith) [reply draft-ietf-rats-corim.md:1230]: WFM

File: draft-ietf-rats-corim.md [hunk 19]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+## Adding DICE/SPDM evidence to the Accepted Claims Set {#sec-dice-spdm}
+
+This section defines how evidence from DICE and/or SPDM is transformed into a
+format where it can be added to an accepted claims set.
+A Verifier supporting DICE/SPDM format evidence should implement this section.
+
+### Transforming SPDM Evidence to a format usable for matching
+
+[Evidence Binding For SPDM](TCG_SPDM-TBD) describes the process by which
+evidence in a SPDM MEASUREMENTS response is converted to Evidence suitable for
```
[the remark this answers is not in the document]
Comment (andrew-draper) [reply draft-ietf-rats-corim.md:1418]: This section is talking about DICE. The SPDM section is later (unless your comment was moved)

File: draft-ietf-rats-corim.md (hunk 18 above)
Comment (andrew-draper) [reply draft-ietf-rats-corim.md:1262]: Using Yogesh's proposal

File: draft-ietf-rats-corim.md (hunk 1 above)
Comment (andrew-draper) [reply draft-ietf-rats-corim.md:1230]: standard algorithm it is

File: draft-ietf-rats-corim.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (andrew-draper) [reply draft-ietf-rats-corim.md:1264]: The verification process could activate tags which are not CoMIDs, so I think referring to tags is appropriate here.

File: draft-ietf-rats-corim.md [hunk 20]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+For example, if the Evidence format is known in advance, tags that do not match
+the expected profile can be readily discarded.
+
+The selection process MUST yield at least one usable tag.
+
+### CoBOM Extraction
+
+All the available Concise Bill Of Material (CoBOMs) tags are then collected
+from the selected CoRIMs.
+
```
[the remark this answers is not in the document]
Comment (andrew-draper) [reply draft-ietf-rats-corim.md:1270]: I added a paragraph in the section above telling the verifier to discard invalid CoBOMs before using them.

File: draft-ietf-rats-corim.md [hunk 21]
```diff
@@ -1207,6 +1207,242 @@ The following describes each member of the `concise-bom-tag` map.
...
+The Verifier MUST activate all tags referenced by a CoBOM.
+
+After the Verifier has processed all CoBOMs it MUST discard any tags which have
+not been activated by a CoBOM.
+
+### Tags Identification and Validation
```
[the remark this answers is not in the document]
Comment (andrew-draper) [reply draft-ietf-rats-corim.md:1276]: I think the two steps are separate, so it is clearer to describe them sequentially. If you feel strongly then we can reconsider.

File: draft-ietf-rats-corim.md (hunk 3 above)
Comment (andrew-draper) [reply draft-ietf-rats-corim.md:1281]: I agree, no need to state that twice.

File: draft-ietf-rats-corim.md (hunk 9 above)
Comment (andrew-draper) [reply draft-ietf-rats-corim.md:1292]: We need to make a decision about whether tag boundaries are used as part of matching. Recent conversations in the group were going in the direction where they are ignored, but I think it may simplifiy things if we keep them. I will turn this into an issue.

File: draft-ietf-rats-corim.md (hunk 13 above)
Comment (andrew-draper) [reply draft-ietf-rats-corim.md:1351]: I agree, this should go into a separate file (and be formatted correctly). Can I do that with a separate change after we accept this one?

**Review by nedmsmith — APPROVED** [review]

**Review by thomas-fossati — APPROVED** [review]

**Review by henkbirkholz — APPROVED** [review]
